### PR TITLE
add oneof support

### PR DESCRIPTION
**Pull request by Razangann**
add oneof support and fix log4j-api version to actual.

### CODE REVIEW

File: src/main/java/com/vizor/unreal/convert/ProtoProcessor.java [hunk 1]
```diff
@@ -15,30 +15,14 @@
...
 package com.vizor.unreal.convert;
 
-import com.squareup.wire.schema.internal.parser.EnumConstantElement;
-import com.squareup.wire.schema.internal.parser.EnumElement;
-import com.squareup.wire.schema.internal.parser.FieldElement;
-import com.squareup.wire.schema.internal.parser.MessageElement;
-import com.squareup.wire.schema.internal.parser.ProtoFileElement;
-import com.squareup.wire.schema.internal.parser.ServiceElement;
-import com.squareup.wire.schema.internal.parser.TypeElement;
+import com.squareup.wire.schema.internal.parser.*;
```
Comment (VizorDaeva) [inline src/main/java/com/vizor/unreal/convert/ProtoProcessor.java]: Уверен что хорошая идея так делать? Так вместо точечных импортов ты тянешь прямо всё (я не знаю на сколько там много этого всего) и это излишние затраты памяти.
Хотя я мыслю с позиции плюсов и может к джаве это и не относится

File: src/main/java/com/vizor/unreal/convert/ProtoProcessor.java [hunk 2]
```diff
@@ -406,6 +394,32 @@ private CppStruct extractStruct(final TypesProvider provider, final MessageEleme
...
+
+            final String fieldName = provider.fixFieldName(onf.name(), ueType.isA(boolean.class));
+            field = new CppField(ueType, fieldName);
+
+            final String sourceDoc = onf.documentation();
+            if (!sourceDoc.isEmpty())
+                field.javaDoc.set(sourceDoc);
+
+            field.enableAnnotations(false);
+            //field.addAnnotation(fieldAnnotations);
```
Comment (VizorDaeva) [inline src/main/java/com/vizor/unreal/convert/ProtoProcessor.java]: Забыл удалить?

File: src/main/java/com/vizor/unreal/convert/ProtoProcessor.java (hunk 1 above)
Comment (Razangann) [reply src/main/java/com/vizor/unreal/convert/ProtoProcessor.java]: Хмм... у меня IDE сам дописывал импорты, я даже не смотрел на них)

File: src/main/java/com/vizor/unreal/convert/ProtoProcessor.java (hunk 2 above)
Comment (Razangann) [reply src/main/java/com/vizor/unreal/convert/ProtoProcessor.java]: Пока не нашел как для  TUnion использовать какой-нибудь UPROPERTY, временно убрал эту строку, думал может кто подскажет и верну аннотацию:)

Comment (VizorDaeva) [reply src/main/java/com/vizor/unreal/convert/ProtoProcessor.java]: К сожалению TUnion не может быть UPROPERTY так что просто удаляй её

**Review by nikomoravec — CHANGES_REQUESTED** [review]
oneof realization via TUnion makes oneof fields inaccessible from blueprints. It goes against the original requirements and ideas of Infraworld project

**Comment by Razangann** [comment]
> oneof realization via TUnion makes oneof fields inaccessible from blueprints. It goes against the original requirements and ideas of Infraworld project

unfortunately TUnion doesn't work with UPTOPERTY. i dont know how to use in blueprints. any suggestions?

**Comment by nikomoravec** [comment]
@Razangann

> unfortunately TUnion doesn't work with UPTOPERTY. i dont know how to use in blueprints. any suggestions?

I think you may try to implement this by creating your own OneOf implementation with TVariant inside in [InfraworldRuntime](https://github.com/vizor-games/InfraworldRuntime) project and adopt Cornerstone then.

**Comment by Razangann** [comment]
> @Razangann
> 
> > unfortunately TUnion doesn't work with UPTOPERTY. i dont know how to use in blueprints. any suggestions?
> 
> I think you may try to implement this by creating your own OneOf implementation with TVariant inside in [InfraworldRuntime](https://github.com/vizor-games/InfraworldRuntime) project and adopt Cornerstone then.
TVaritan also cannot handle UPROPERTY.

**Comment by Razangann** [comment]
> @Razangann
> 
> > unfortunately TUnion doesn't work with UPTOPERTY. i dont know how to use in blueprints. any suggestions?
> 
> I think you may try to implement this by creating your own OneOf implementation with TVariant inside in [InfraworldRuntime](https://github.com/vizor-games/InfraworldRuntime) project and adopt Cornerstone then.

TVaritan also cannot handle UPROPERTY.

**Comment by Razangann** [comment]
> @Razangann
> 
> > unfortunately TUnion doesn't work with UPTOPERTY. i dont know how to use in blueprints. any suggestions?
> 
> I think you may try to implement this by creating your own OneOf implementation with TVariant inside in [InfraworldRuntime](https://github.com/vizor-games/InfraworldRuntime) project and adopt Cornerstone then.

TVaritan also cannot handle UPROPERTY.

**Comment by nikomoravec** [comment]
> TVaritan also cannot handle UPROPERTY.

Yes, it cant. But as I said above you can implement your own OneOf structure with TVariant inside it.

Something like:

``USTRUCT()
struct OneOf
{
  GENERATED_BODY()
  // empty stub structrure
};``


Image that you have oneof fileld in you proto file:


```
message SampleMessage {
  oneof my_oneof {
    string oneofString = 1;
    int32 oneofInteger = 2;
    float oneOfFloat = 3;
  }
}
```

in this case you  need to generate something like this:

```
USTRUCT()
SampleMessageOneOf : public OneOf
{
  GENERATED_BODY()
public:
  UFUNCTION(BlueprintCallable)
  FString GetStringValue()
  {
    return Variant.Get<FString>();
  }

  UFUNCTION(BlueprintCallable)
  int32 GetInt32Value()
  {
    return Variant.Get<int32>();
  }

  UFUNCTION(BlueprintCallable)
  float GetFloatValue()
  {
    return Variant.Get<float>();
  }

private:
  TVariant<FString, int32, float> Variant;
};

**Comment by Razangann** [comment]
> > TVaritan also cannot handle UPROPERTY.
> 
> Yes, it cant. But as I said above you can implement your own OneOf structure with TVariant inside it.
> 
> Something like:
> 
> `USTRUCT() struct OneOf { GENERATED_BODY() // empty stub structrure };`
> 
> Image that you have oneof fileld in you proto file:
> 
> ```
> message SampleMessage {
>   oneof my_oneof {
>     string oneofString = 1;
>     int32 oneofInteger = 2;
>     float oneOfFloat = 3;
>   }
> }
> ```
> 
> in this case you need to generate something like this:
> 
> ```
> USTRUCT()
> SampleMessageOneOf : public OneOf
> {
>   GENERATED_BODY()
> public:
>   UFUNCTION(BlueprintCallable)
>   FString GetStringValue()
>   {
>     return Variant.Get<FString>();
>   }
> 
>   UFUNCTION(BlueprintCallable)
>   int32 GetInt32Value()
>   {
>     return Variant.Get<int32>();
>   }
> 
>   UFUNCTION(BlueprintCallable)
>   float GetFloatValue()
>   {
>     return Variant.Get<float>();
>   }
> 
> private:
>   TVariant<FString, int32, float> Variant;
> };
> ```

USTRUCT cannot contain UFUNCTION inside. Also I thought to use USCLASS with UFUNCTION and add to USTRUCT. but the class inside the struct is also not used

**Pull request merged**